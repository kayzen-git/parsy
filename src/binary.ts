import {ParsyContext, ParsyParser} from '.';

export abstract class BinaryParser extends ParsyParser {
    public left!: ParsyParser;
    public right!: ParsyParser;

    public set(left: ParsyParser, right: ParsyParser): this {
        this.left = left;
        this.right = right;
        return this;
    }
}

export class AndParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        return this.left.parse(context) && this.right.parse(context);
    }
}

export class OrParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        const {index} = context;
        if (this.left.parse(context)) {
            return context;
        } else {
            context.index = index;
            return this.right.parse(context);
        }
    }
}
