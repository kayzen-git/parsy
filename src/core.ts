import { ParsyParser } from './index';

export function terminal<I, O>(
    evaluate: (context: I) => Iterable<O> | null
): ParsyParser<I, O> {
    return function* (context) {
        const value = evaluate(context);
        if (value === null) {
            return false;
        } else {
            yield* value;
            return true;
        }
    };
}

export function and<I, O>(
    left: ParsyParser<I, O>,
    right: ParsyParser<I, O>
): ParsyParser<I, O> {
    return function* (context) {
        return (yield* left(context)) && (yield* right(context));
    };
}

export function or<I, O>(
    left: ParsyParser<I, O>,
    right: ParsyParser<I, O>
): ParsyParser<I, O> {
    return function* (context) {
        return (yield* left(context)) || (yield* right(context));
    };
}
