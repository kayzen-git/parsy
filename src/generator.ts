export function char([value]: string) {
    return function* (tokens: Iterable<string>): Generator<string | null> {
        for (const token of tokens) {
            yield token === value ? token : null;
        }
    };
}

for (const token of char('a')('aaa')) {
    console.log('asd');
}

export function sequence(...parsers: Array<IterableIterator<any>>) {
    return function* (context: IterableIterator<any>) {
        const pattern = parsers.map((parser) => parser(context));
        for (const parser of pattern) {
            for (const token of parser) {
                yield token;
            }
        }
    };
}

export function alternation(...parsers) {
    return function* (context) {
        const pattern = parsers.map((parser) => parser(context));
        const map = new Map();
        for (const parser of pattern) {
            for (const token of parser) {
                yield [];
            }
        }
    };
}

export function and<T>(...parsers: Array<Generator<T | null>>) {
    return function* (tokens: IterableIterator<any>) {
        const pattern = parsers.map((parser) => parser(tokens));
        for (const parser of pattern) {
            yield* parser;
        }
    };
}

function* advance(tokens) {
    for (const token of tokens) {
        while (!(yield token));
    }
}

export function* or<T>(...parsers: Array<Generator<T | null>>) {
    return function* (tokens) {
        // const stream = advance(tokens);
        for (const token of tokens) {
            for (const parser of parsers) {
                yield parser.next(token);
            }
        }
    };
}

export function* aggregate(tokens) {
    for (const token of tokens) {
        yield [...token];
    }
}

export function join(...parsers) {
    return (tokens) => [[...sequence(...parsers)(tokens)].join('')];
}

interface ParsyStream<T> extends Iterator<T> {
    peek(...args): T;
}

// export function stream<T>(iter: Iterator<T>): ParsyStream<T> {
//     let value: any;
//     return {
//         ...iter,
//         next(...args: any) {
//             if (value === undefined) {
//                 return iter.next(...args);
//             } else {
//                 const temp = value;
//                 value = undefined;
//                 return temp;
//             }
//         },
//         peek(...args: any): any {
//             if (value === undefined) {
//                 return value = iter.next(...args);
//             } else {
//                 const temp = value;
//                 value = undefined;
//                 return temp;
//             }
//         },
//     }
// }

export function parsy(grammar: GeneratorFunction) {
    return function (input: string, index = 0) {
        return [...grammar(input.substring(index)[Symbol.iterator]())];
    };
}
