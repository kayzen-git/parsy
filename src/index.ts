export * from './core';
export * from './terminals';
export * from './parsy';

export type ParsyParser<I = any, O = any> = (
    input: I
) => Generator<O, boolean, undefined>;
export type ParsyTransformer<I, T> = (token: I) => T;

export interface ParsyContext {
    input: string;
    index: number;
}

export interface ParsyToken<T = any> extends ParsyContext {
    type: string;
    value: string;
    length: number;
    tokens: T[];
}
