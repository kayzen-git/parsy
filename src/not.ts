import {ParsyContext, ParsyParser} from './index';

export class NotParser extends ParsyParser {
    public parser!: ParsyParser;

    public set(parser: ParsyParser): this {
        this.parser = parser;
        return this;
    }

    public parse(context: ParsyContext): ParsyContext | undefined {
        if (this.parser.parse(context)) {
            return undefined;
        } else {
            return context.advance(1);
        }
    }
}
