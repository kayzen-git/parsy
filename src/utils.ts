import { ParsyParser } from './index';

export function collect<T>(
    iterator: IterableIterator<T>,
    listen?: (value: T) => void
): Array<T> | null {
    const values = new Array<T>();
    let result = iterator.next();
    while (!result.done) {
        listen && listen(result.value);
        values.push(result.value);
        result = iterator.next();
    }
    return result.value ? values : null;
}

export function split(...parsers: ParsyParser[]): ParsyParser {
    return function* (context) {
        const pattern = parsers.map((parser) => parser(context));
        while (parsers.length > 1) {
            yield pattern.map((parser) => parser.next());
        }
        return yield* pattern[0];
    };
}

export function assign(
    helper: (...args: any) => (...args: any) => any,
    ...args
) {
    const target = (context) => helper(...args)(context);
    return Object.assign(target, {
        set(...values) {
            args = values;
            return target;
        },
    });
}
