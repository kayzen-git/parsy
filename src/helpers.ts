import {AndParser, OrParser, ParsyParser, ParsyToken, RuleParser, TerminalParser, TransformParser} from '.';
import {NotParser} from './not';

/**
 * Matches the empty string.
 */
export const empty = new ParsyParser();

export function or(left: ParsyParser, right: ParsyParser) {
    return new OrParser().set(left, right);
}

export function and(left: ParsyParser, right: ParsyParser) {
    return new AndParser().set(left, right);
}

/**
 * Matches a range of characters.
 */
export function terminal(...pattern: string[]) {
    return new TerminalParser().set(...pattern);
}

/**
 * Returns a list of characters between the given char codes.
 */
export function charset(from: string, to: string) {
    return [...Array(to.charCodeAt(0) - from.charCodeAt(0) + 1).keys()]
        .map(value => String.fromCharCode(value + from.charCodeAt(0)));
}

/**
 * Scopes the passed parser with a type.
 */
export function rule(type: string) {
    return new RuleParser(type);
}

/**
 * Matches the supplied sequence but recovers if not all of the rules match.
 * Makes the sequence optional.
 */
export function optional(parser: ParsyParser) {
    return new OrParser().set(parser, empty);
}

/**
 * Repeats matching the supplied rule for as long as the input allows for it.
 * Must match at least once.
 */
export function repeated(parser: ParsyParser) {
    const self = new AndParser();
    return self.set(parser, optional(self));
}

/**
 * Matches a sequence of rules.
 * All the rules provided need to match.
 */
export function sequence(...parsers: ParsyParser[]) {
    const reducer = (parser, rule) => new AndParser().set(rule, parser);
    return parsers.reduceRight(reducer);
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the rules has to match.
 */
export function alternation(...parsers: ParsyParser[]) {
    const start = new OrParser().set(parsers[parsers.length - 2], parsers[parsers.length - 1]);
    const reducer = (parser, rule) => new OrParser().set(rule, parser);
    return parsers.slice(0, -2).reduceRight(reducer, start);
}

/**
 * Inverts the specified parser's output.
 */
export function not(parser: ParsyParser) {
    return new NotParser().set(parser);
}

/**
 * Listens to the creation of tokens created by the passed parser.
 */
export function transform(parser: ParsyParser, transformer: (token: ParsyToken) => any) {
    return new TransformParser(parser, transformer);
}
