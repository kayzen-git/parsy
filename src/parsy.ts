import {
    AndParser,
    OrParser,
    ParsyContext,
    ParsyParser,
    ParsyTransform,
    RuleParser,
    TerminalParser,
    TransformParser,
} from '.';
import { NotParser } from './not';

export class Parsy {
    constructor(public parser: ParsyParser) {}

    public parse(input: string) {
        const context = new ParsyContext(input);
        const result = this.parser.parse(context);
        if (result) {
            return result.token.children;
        }
    }

    /**
     * Matches the empty string.
     */
    static empty = new ParsyParser();

    static or(left: ParsyParser, right: ParsyParser) {
        return new OrParser().set(left, right);
    }

    static and(left: ParsyParser, right: ParsyParser) {
        return new AndParser().set(left, right);
    }

    /**
     * Matches a range of characters.
     */
    static terminal(...pattern: string[]) {
        return new TerminalParser().set(...pattern);
    }

    /**
     * Returns a list of characters between the given char codes.
     */
    static charset(from: string, to: string) {
        return [...Array(to.charCodeAt(0) - from.charCodeAt(0) + 1).keys()]
            .map((value) => String.fromCharCode(value + from.charCodeAt(0)));
    }

    /**
     * Scopes the passed parser with a type.
     */
    static rule(type: string) {
        return new RuleParser(type);
    }

    /**
     * Matches the supplied sequence but recovers if not all of the rules match.
     * Makes the sequence optional.
     */
    static optional(parser: ParsyParser) {
        return new OrParser().set(parser, this.empty);
    }

    /**
     * Repeats matching the supplied rule for as long as the input allows for it.
     * Must match at least once.
     */
    static repeated(parser: ParsyParser) {
        const self = new AndParser();
        return self.set(parser, this.optional(self));
    }

    /**
     * Matches a sequence of rules.
     * All the rules provided need to match.
     */
    static sequence(...parsers: ParsyParser[]) {
        const reducer = (parser, rule) => new AndParser().set(rule, parser);
        return parsers.reduceRight(reducer);
    }

    /**
     * Searches for the first type that matches in sequential order.
     * At least one of the rules has to match.
     */
    static alternation(...parsers: ParsyParser[]) {
        const start = new OrParser().set(
            parsers[parsers.length - 2],
            parsers[parsers.length - 1]
        );
        const reducer = (parser, rule) => new OrParser().set(rule, parser);
        return parsers.slice(0, -2).reduceRight(reducer, start);
    }

    /**
     * Inverts the specified parser's output.
     */
    static not(parser: ParsyParser) {
        return new NotParser().set(parser);
    }

    /**
     * Listens to the creation of tokens created by the passed parser.
     */
    static transform(parser: ParsyParser, transformer: ParsyTransform) {
        return new TransformParser(parser, transformer);
    }
}
