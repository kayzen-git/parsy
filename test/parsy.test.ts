import {color} from 'ansi-styles';
import {alternation, charset, optional, repeated, rule, sequence, terminal, transform} from '../src';
import {Parsy} from '../src/parser';
import {pretty} from './utils';


describe('parsy', function () {
    const _ = rule('whitespace');
    const IDENTIFIER = rule('identifier');
    const STRING = rule('string');
    const PATTERN = rule('pattern');
    const MODIFIER = rule('modifier');
    const ANY = rule('any');
    const RULE = rule('rule');
    const SEQUENCE = rule('sequence');
    const ALTERNATION = rule('alternation');
    const EXPRESSION = rule('expression');

    _.set(repeated(terminal(...' \n\t\r')));
    IDENTIFIER.set(repeated(terminal(...charset('a', 'z'), ...charset('A', 'Z'), ...charset('0', '9'))));
    // STRING.set(sequence(terminal('\''), optional(repeated(alternation(terminal('\\\''), terminal(...charset(String.fromCharCode(33), String.fromCharCode(126)))))), terminal('\'')));
    PATTERN.set(sequence(terminal('['), IDENTIFIER, terminal(']')));
    MODIFIER.set(terminal(...'*+?'));
    RULE.set(sequence(IDENTIFIER, optional(_), terminal(':'), optional(_), EXPRESSION));
    SEQUENCE.set(sequence(terminal('('), optional(_), repeated(EXPRESSION), optional(_), terminal(')')));
    ALTERNATION.set(sequence(terminal('{'), optional(_), repeated(EXPRESSION), optional(_), terminal('}')));
    EXPRESSION.set(sequence(alternation(SEQUENCE, ALTERNATION, ANY, IDENTIFIER), optional(MODIFIER)));

    const parser = new Parsy(sequence(optional(_), repeated(RULE), optional(_)));
    const input = `
        start: rule+
        identifier: [A-Za-z0-9]+
        string: ( ['] { '\\\\\\'' . }* ['] )
        pattern: ( '[' .* ']' )
        modifier: [*+?]
        any: '.'
        rule: ( identifier ':' expression )
        sequence: ( '(' expression+ ')' )
        alternation: ( '{' expression+ '}' )
        expression: ( { sequence alternation string any identifier } modifier? )
    `;

    // it('should parse correct structure', function () {
    //     expect(pretty(parser.parse(input))).toMatchSnapshot();
    // });
});

describe('scopy', function () {
    const LABEL = transform(rule('label'), token => token.value = color.blue.open + token.value + color.blue.close);
    const OPEN = rule('open');
    const CLOSE = rule('close');
    const VALUE = rule('value');

    LABEL.set(repeated(terminal(...charset('a', 'z'), ...charset('A', 'Z'), ...charset('0', '9'))));
    OPEN.set(terminal('{'));
    CLOSE.set(terminal('}'));
    VALUE.set(sequence(LABEL, optional(sequence(OPEN, optional(repeated(VALUE)), CLOSE))));

    const parser = new Parsy(VALUE);
    const input = 'scope{another{one{}}like{this}}';

    it('should parse correct structure', function () {
        expect(pretty(parser.parse(input)![0])).toMatchSnapshot();
    });
});

describe('math', function () {
    const INTEGER = transform(rule('integer'), token => token.value = color.blueBright.open + token.value + color.blueBright.close);
    // const VARIABLE = rule('variable');
    // const LINE = rule('line');
    // const DOT = rule('dot');
    // const EXPRESSION = rule('expression');
    // const TERM = rule('term');
    // const FACTOR = rule('factor');
    const PRIMARY = rule('primary');
    const MULTIPLICATIVE = rule('multiplicative');
    const ADDITIVE = rule('additive');
    const ADDITION = rule('operator').set(terminal('+'));
    const SUBTRACTION = rule('operator').set(terminal('-'));
    const MULTIPLICATION = rule('operator').set(terminal('*'));
    const DIVISION = rule('operator').set(terminal('/'));
    const OPEN = rule('open').set(terminal('('));
    const CLOSE = rule('close').set(terminal(')'));

    INTEGER.set(repeated(terminal(...charset('0', '9'))));
    ADDITIVE.set(alternation(sequence(MULTIPLICATIVE, alternation(ADDITION, SUBTRACTION), ADDITIVE), MULTIPLICATIVE));
    MULTIPLICATIVE.set(alternation(sequence(PRIMARY, alternation(MULTIPLICATION, DIVISION), MULTIPLICATIVE), PRIMARY));
    PRIMARY.set(alternation(sequence(OPEN, ADDITIVE, CLOSE), INTEGER));

    // INTEGER.set(repeated(terminal(...charset('0', '9'))));
    // VARIABLE.set(repeated(terminal(...charset('a', 'z'), ...charset('A', 'Z'), ...charset('0', '9'))));
    // LINE.set(terminal('+', '-'));
    // DOT.set(terminal('*', '/'));
    // EXPRESSION.set(alternation(TERM, sequence(EXPRESSION, LINE, TERM), sequence(LINE, TERM)));
    // TERM.set(alternation(FACTOR, sequence(TERM, DOT, FACTOR)));
    // FACTOR.set(alternation(PRIMARY, sequence(FACTOR, terminal('^'), PRIMARY)));
    // PRIMARY.set(alternation(INTEGER, sequence(terminal('('), EXPRESSION, terminal(')'))));

    const parser = new Parsy(ADDITIVE);
    const input = '2-2*((10+0)+100000000000000000000000)';

    it('should parse correct structure', function () {
        expect(pretty(parser.parse(input))).toMatchSnapshot();
    });
});
