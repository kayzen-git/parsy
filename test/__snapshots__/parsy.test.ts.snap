// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`math should parse correct structure 1`] = `
"
Array [
  additive {
    multiplicative {
        primary {
            integer: '[94m2[39m'
        }
    }
    operator: '-'
    additive {
        multiplicative {
            primary {
                integer: '[94m2[39m'
            }
            operator: '*'
            multiplicative {
                primary {
                    open: '('
                    additive {
                        multiplicative {
                            primary {
                                open: '('
                                additive {
                                    multiplicative {
                                        primary {
                                            integer: '[94m10[39m'
                                        }
                                    }
                                    operator: '+'
                                    additive {
                                        multiplicative {
                                            primary {
                                                integer: '[94m0[39m'
                                            }
                                        }
                                    }
                                }
                                close: ')'
                            }
                        }
                        operator: '+'
                        additive {
                            multiplicative {
                                primary {
                                    integer: '[94m100000000000000000000000[39m'
                                }
                            }
                        }
                    }
                    close: ')'
                }
            }
        }
    }
},
]
"
`;

exports[`scopy should parse correct structure 1`] = `
"
value {
  label: '[34mscope[39m'
  open: '{'
  value {
    label: '[34manother[39m'
    open: '{'
    value {
      label: '[34mone[39m'
      open: '{'
      close: '}'
    }
    close: '}'
  }
  value {
    label: '[34mlike[39m'
    open: '{'
    value {
      label: '[34mthis[39m'
    }
    close: '}'
  }
  close: '}'
}
"
`;
